packages/flutter_tools/gradle/src/main/kotlin/DependencyVersionChecker.kt

```diff
@@ -0,0 +1,331 @@
+
+    @VisibleForTesting internal val warnAGPVersion: AndroidPluginVersion = AndroidPluginVersion(7, 3, 1)
+
+    @VisibleForTesting internal val errorAGPVersion: AndroidPluginVersion = AndroidPluginVersion(7, 0, 0)
+
+    @VisibleForTesting internal val warnKGPVersion: Version = Version(1, 8, 10)
+
+    @VisibleForTesting internal val errorKGPVersion: Version = Version(1, 7, 0)
+
+    /**
+     * Checks if the project's Android build time dependencies are each within the respective
+     * version range that we support. When we can't find a version for a given dependency
+     * we treat it as within the range for the purpose of this check.
+     */
+    @JvmStatic fun checkDependencyVersions(project: Project) {
+        project.extra.set(OUT_OF_SUPPORT_RANGE_PROPERTY, false)
+
+        checkGradleVersion(getGradleVersion(project), project)
+        checkJavaVersion(getJavaVersion(), project)
+        val agpVersion: AndroidPluginVersion? = getAGPVersion(project)
+        if (agpVersion != null) {
+            checkAGPVersion(agpVersion, project)
+        } else {
+            project.logger.error(
+                "Warning: unable to detect project AGP version. Skipping " +
+                    "version checking. \nThis may be because you have applied AGP after the Flutter Gradle Plugin."
+            )
+        }
+
+        val kgpVersion: Version? = getKGPVersion(project)
+        if (kgpVersion != null) {
+            checkKGPVersion(kgpVersion, project)
+        }
+        // KGP is not required, so don't log any warning if we can't find the version.
+    }
+
+    // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.invocation/-gradle/index.html#-837060600%2FFunctions%2F-1793262594
+    @VisibleForTesting internal fun getGradleVersion(project: Project): Version {
+        val untrimmedGradleVersion: String = project.gradle.gradleVersion
+        // Trim to handle candidate gradle versions (example 7.6-rc-4). This means we treat all
+        // candidate versions of gradle as the same as their base version
+        // (i.e., "7.6"="7.6-rc-4").
+        return Version.fromString(untrimmedGradleVersion.substringBefore('-'))
+    }
+
+    // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api/-java-version/index.html#-1790786897%2FFunctions%2F-1793262594
+    @VisibleForTesting internal fun getJavaVersion(): JavaVersion {
+        return JavaVersion.current()
+    }
+
+    @VisibleForTesting internal fun getAGPVersion(project: Project): AndroidPluginVersion? {
+        val androidPluginVersion: AndroidPluginVersion? =
+            project.extensions.findByType(
+                AndroidComponentsExtension::class.java
+            )?.pluginVersion
+        return androidPluginVersion
+    }
+
+    // TODO(gmackall): AGP has a getKotlinAndroidPluginVersion(), and KGP has a
+    //                 getKotlinPluginVersion(). Consider replacing this implementation with one of
+    //                 those.
```

Comment:
Fantastic find!

Reply:
FWIW I think there is also a AndroidPluginVersion https://developer.android.com/reference/tools/gradle-api/8.8/com/android/build/api/AndroidPluginVersion#getCurrent()